clang/test/SemaSYCL/sycl-external-attr.cpp

```diff
@@ -0,0 +1,70 @@
+// RUN: %clang_cc1 -fsycl-is-device -fsyntax-only -verify %s
+// RUN: %clang_cc1 -fsycl-is-device -std=c++20 -fsyntax-only -verify -DCPP20 %s
+// Semantic tests for sycl_external attribute
+
+[[clang::sycl_external]] // expected-error {{'sycl_external' can only be applied to functions with external linkage}}
+static void func1() {}
+
+namespace {
+  [[clang::sycl_external]] // expected-error {{'sycl_external' can only be applied to functions with external linkage}}
+  void func2() {}
+
+  struct UnnX {};
+}
+
+[[clang::sycl_external]] // expected-error {{'sycl_external' can only be applied to functions with external linkage}}
+  void func4(UnnX) {}
+
+// The first declaration of a SYCL external function is required to have this attribute.
+int foo(); // expected-note {{previous declaration is here}}
+
+[[clang::sycl_external]] int foo(); // expected-error {{'clang::sycl_external' attribute does not appear on the first declaration}}
+
+// Subsequent declrations of a SYCL external function may optionally specify this attribute.
+[[clang::sycl_external]] int boo();
+
+[[clang::sycl_external]] int boo(); // OK
+
+int boo(); // OK
+
+class C {
+  [[clang::sycl_external]] void member();
+};
+
+[[clang::sycl_external]] int main() // expected-error {{'sycl_external' cannot be applied to the 'main' function}}
+{
+  return 0;
+}
+
+class D {
+  [[clang::sycl_external]] void del() = delete; // expected-error {{'sycl_external' cannot be applied to an explicitly deleted function}}
+};
```

Comment:
This test doesn't require a member function, but the presence of class `D` suggests that there is something specific to member functions here.
```suggestion
// expected-error {{'sycl_external' cannot be applied to an explicitly deleted function}}
[[clang::sycl_external]] void del() = delete;
```

Reply:
For the `sycl_kernel_entry_point` attribute, I had to add some checks, including the one for a deleted function, directly to `Sema::ActOnFinishFunctionBody`. See https://github.com/llvm/llvm-project/blob/711132dfa43211bce312db01a473aed51bae7b8c/clang/lib/Sema/SemaDecl.cpp#L16246-L16276. In my case, I had to check `isDeleted()`, but your existing code is correct in checking for `isDeletedAsWritten()`.

Reply:
Thank you! That fixes this issue.